cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -160,18 +160,34 @@ Status ResolveOneFieldRef(
     const std::unordered_map<std::string, const SchemaField*>& field_lookup,
     const std::unordered_set<std::string>& duplicate_fields,
     std::vector<int>* columns_selection) {
-  if (const std::string* name = field_ref.name()) {
+  auto resolve_field_ref = [&](const std::string* name) -> Status {
     auto it = field_lookup.find(*name);
     if (it != field_lookup.end()) {
       AddColumnIndices(*it->second, columns_selection);
     } else if (duplicate_fields.find(*name) != duplicate_fields.end()) {
-      // We shouldn't generally get here because SetProjection will reject such references
+      // We shouldn't generally get here because SetProjection will reject such
+      // references
       return Status::Invalid("Ambiguous reference to column '", *name,
                              "' which occurs more than once");
     }
     // "Virtual" column: field is not in file but is in the ScanOptions.
     // Ignore it here, as projection will pad the batch with a null column.
     return Status::OK();
+  };
+
+  if (const std::string* name = field_ref.name()) {
+    return resolve_field_ref(name);
+  } else if (const FieldPath* path = field_ref.field_path()) {
+    auto indices = path->indices();
+    if (indices.size() > 1) {
+      return Status::NotImplemented("Provided FieldRef ", field_ref.ToString(),
+                                    ", Nested FieldPaths are not supported!");
+    }
+    int index = indices[0];
+    auto schema_field = manifest.schema_fields[index];
```

Comment:
`manifest.schema_fields` will give us the fields in file order.  But, an integral field reference is actually a reference into the dataset schema.  So I think we want to be using the dataset schema here instead of the file manifest.

We should add a test case for this as it is a rather subtle point...

```
Parquet File 1:
x | y | z
1 | 2 | 3
```

```
Parquet File 2:
z | y | x
3 | 2 | 1
```

```
Dataset Schema:
 - x: int32
 - y: int32
 - z: int32
```

`field_ref(0)` should yield `[1, 1]`

Reply:
Thanks you for noting this one.